monitoring/interpod.go

```diff
@@ -229,28 +229,50 @@ func waitTimeoutForPodRunningInNamespace(client *kube.Clientset, podName string,
 
 // waitForPodCondition waits until a pod is in the given condition within the specified amount of time.
 func waitForPodCondition(client *kube.Clientset, ns, podName, desc string, timeout time.Duration, condition podCondition) error {
-	log.Infof("waiting up to %v for pod %s status to be %s", timeout, podName, desc)
-	for start := time.Now(); time.Since(start) < timeout; time.Sleep(pollInterval) {
-		pod, err := client.CoreV1().Pods(ns).Get(podName, metav1.GetOptions{})
-		if err != nil {
-			log.Infof("get pod %s in namespace '%s' failed, ignoring for %v: %v",
-				podName, ns, pollInterval, err)
-			continue
-		}
-		done, err := condition(pod)
-		if done {
-			// TODO: update to latest trace to wrap nil
+	log := log.WithFields(log.Fields{
+		"pod":       podName,
+		"namespace": ns,
+	})
+
+	log.Infof("waiting up to %v for pod status to be %s", timeout, desc)
+
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	ticker := time.NewTicker(pollInterval)
+	defer ticker.Stop()
+
+	start := time.Now()
+
+	for {
+		select {
+		case <-ticker.C:
+			pod, err := client.CoreV1().Pods(ns).Get(ctx, podName, metav1.GetOptions{})
+			if err != nil {
+				log.WithError(err).
+					Infof("failed to get pod, retrying after %v", pollInterval)
+				continue
+			}
+
+			done, err := condition(pod)
+			if !done {
+				log.Infof("waiting for pod status to be '%s' (found phase: %q, readiness: %t) (%v elapsed)",
+					desc, pod.Status.Phase, podReady(pod), time.Since(start))
+				continue
+			}
+
 			if err != nil {
```

Comment:
Wouldn't the code will basically never reach inside this "err != nil" block because presumably if err != nil than "done" will be false and thus it will always do "continue" above? Maybe you want to check "err != nil" first and then "!done"?

Reply:
I just kept the same code paths as before and didn't bother to look at the implementation of `condition` 😅 , but it looks like the only time `err != nil` is when `done == true`. On line 218:
```golang
        return waitForPodCondition(client, namespace, podName, "running", timeout, func(pod *v1.Pod) (bool, error) {
		if pod.Status.Phase == v1.PodRunning {
			log.Infof("found pod '%s' on node '%s'", podName, pod.Spec.NodeName)
			return true, nil
		}
		if pod.Status.Phase == v1.PodFailed {
			return true, trace.Errorf("pod in failed status: %s", fmt.Sprintf("%#v", pod))
		}
		return false, nil
	})
```